tango/__main__.py

```diff
@@ -656,9 +732,48 @@ def _run(
                     + click.style(f"{info.result_location}", bold=True, fg="green")
                 )
 
-        click_logger.info(
-            click.style("Finished run ", fg="green") + click.style(run.name, fg="green", bold=True)
-        )
+        if step_name is not None:
+            uncacheable_leaf_steps = step_graph.uncacheable_leaf_steps()
+            step = step_graph[step_name]
+            executor.execute_step(step, step_name in uncacheable_leaf_steps)
+            log_step_summary(step)
+            click_logger.info(
+                click.style("Finished run for step ", fg="green")
+                + click.style(step_name, fg="green", bold=True)
+                + click.style(" (", fg="green", bold=True)
+                + click.style(run.name, fg="green", bold=True)
+                + click.style(")", fg="green", bold=True)
+            )
+        else:
+            failed_steps = executor.execute_step_graph(step_graph, run_name=run.name)
+            # Print everything that has been computed.
+            ordered_steps = sorted(step_graph.values(), key=lambda step: step.name)
+            for step in ordered_steps:
+                if step.name in failed_steps:
+                    # TODO: add needed_by
+                    click_logger.info(
+                        click.style(f'\N{cross mark} "{step.name}" failed.', fg="red")
+                    )
+                else:
+                    log_step_summary(step)
+
+            click_logger.info(
+                click.style("Finished run ", fg="green")
+                + click.style(run.name, fg="green", bold=True)
+            )
+
+    if called_by_executor:
+        from tango.common.aliases import EnvVarNames
+
+        # We set this environment variable so that any steps that contain multiprocessing
+        # and call `initialize_worker_logging` also log the messages with the `step_name` prefix.
+        os.environ[EnvVarNames.LOGGING_PREFIX.value] = f"step {step_name}"
+        initialize_prefix_logging(f"step {step_name}", main_process=False)
+        log_and_execute_run()
```

Comment:
Can we use this method of initializing logging all the time, instead of as a one-off for `MulticoreExecutor`?

I don't have the overview over all the logging scenarios, so take my opinions on the matter with a grain of salt. All I see is some code smells in the `called_by_executor` flag, the global state, and the one-off code path. I fear we're fairly committed to doing opaque env var magic at this point. We already do that in other cases. But do we have the one-off code paths? Can we just use the prefix stuff all the time?

Reply:
For the single process executor, there's no point in calling `initialize_prefix_logging`, as the logging has already been initialized once, for the main process (and it is only being called for `step_name`; which we won't have in the general case). I don't particularly like the use of global env variables myself, but we already do it for the rest of the multiprocessing-friendly logging. 

The reason why I added `called_by_executor` was because I wanted to accommodate the case for people running a single step themselves without it necessarily being part of the `MulticoreExecutor` run. 

There were some issues related to sockets if I tried to just run `initialize_prefix_logging` all the time. I can spend some more time on it.